Conditional writes are incorrectly handled in Array allocation sinking

kmiller68 opened the pull request:
#### 934b1e28a87a1a9dbd5011751c819d56f6229734
<pre>
Conditional writes are incorrectly handled in Array allocation sinking
<a href="https://bugs.webkit.org/show_bug.cgi?id=299956">https://bugs.webkit.org/show_bug.cgi?id=299956</a>
<a href="https://rdar.apple.com/161681941">rdar://161681941</a>

Reviewed by Yusuke Suzuki and Yijia Huang.

The current bottom value in ObjectAllocationSinking is incorrect for arrays.
Unlike with objects, which track conditional stores by passing the active
structure through SSA, arrays can&apos;t do this. Instead we should set default value
to the appropriate hole value for the given IndexingShape. To make this work
I had to fix some Phi/Upsilon ResultFormat bugs since they previously assumed
everything would be a JSValue.

Also, add ASSERT to FTL lowering that the Phi/Upsilon formats match. I spent 1/2 a day
trying to understand why I was getting zero, when the issue was those values disagreed
and I was getting the default zero value.

Tests: JSTests/stress/array-allocation-sink-conditional-write-osr.js
       JSTests/stress/array-sink-materialize-conditional-write-argument-value.js
       JSTests/stress/array-sink-materialize-conditional-write.js

Canonical link: <a href="https://commits.webkit.org/300888@main">https://commits.webkit.org/300888@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ceb383a9b4da2ab9950990cc194ec5a061a88f57

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124057 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43767 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34469 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130885 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76199 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4bcfcc67-3753-438f-9521-06700b831716) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44496 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52363 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94373 "23 flakes 26 failures") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62615 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/33818b4d-c78c-421b-886c-49351cb0b0fd) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127011 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35456 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110976 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74969 "Found 1 new API test failure: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlotTimeout (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ebea0d94-6c13-4119-96fd-57bf61a8474f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34404 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29142 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74368 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116211 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105190 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29359 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133557 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/122589 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51004 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38856 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102840 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51380 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107195 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102654 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26144 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48002 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26254 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47882 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50859 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56626 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/153683 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50328 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39061 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53674 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52002 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -1985,8 +1985,27 @@ class ObjectAllocationSinkingPhase : public Phase {
         UncheckedKeyHashMap<FrozenValue*, Node*> lazyMapping;
-        if (!m_bottom)
-            m_bottom = m_insertionSet.insertConstant(0, m_graph.block(0)->at(0)->origin, jsNumber(1927));
+        m_bottom = m_insertionSet.insertConstant(0, m_graph.block(0)->at(0)->origin, jsNumber(1927));
```
Why remove the conditional init for `m_bottom`? Are we inserting a new bottom value here every time?

kmiller68 replied:
This is the first time m_bottom is reached so it was always set here.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -2914,6 +2942,8 @@ class ObjectAllocationSinkingPhase : public Phase {
         }
     }
 
+    // FIXME: Is this needed? We generate the Phi for each value before adding Upsilons, so it seems like we could do this
+    // when inserting the Upsilons.
```
Right, I think we should do that when inserting the Upsilons.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -1985,8 +1985,27 @@ class ObjectAllocationSinkingPhase : public Phase {
-            m_bottom = m_insertionSet.insertConstant(0, m_graph.block(0)->at(0)->origin, jsNumber(1927));
+        m_bottom = m_insertionSet.insertConstant(0, m_graph.block(0)->at(0)->origin, jsNumber(1927));
+
+        auto defaultFor = [&](PromotedHeapLocation location) -> Node* {
+            // For objects we track the transition state of the object (i.e. which properties are conditionally initialized).
+            // Arrays have no such mechanism however they do have a hole value which we can use instead. This is fine because
+            // if the hole conditionally makes it to a materialization it will be stored to the slot and it will appear as
+            // if the slot was never written to.
+            if (location.kind() == ArrayIndexedPropertyPLoc) {
+                ASSERT(location.base()->op() == NewArrayWithButterfly);
+                if (hasDouble(location.base()->indexingType())) {
+                    if (!m_PNaN)
+                        m_PNaN = m_insertionSet.insertConstant(0, m_graph.block(0)->at(0)->origin, jsNumber(PNaN), DoubleConstant);
+                    return m_PNaN;
+                }
+                if (!m_empty)
+                    m_empty = m_insertionSet.insertConstant(0, m_graph.block(0)->at(0)->origin, JSValue());
+                return m_empty;
+            }
+            return m_bottom;
+        };
```
What if an array slot is explicitly initialized first, and then conditionally overwritten? Will the hole value incorrectly replace the initialized value?

hyjorc1 replied:
Like in this test: unconditionally init one slot + conditionally overwrite the SAME  slot.
```
function testConditionalOverwrite(overwrite, escape) {
      let arr = new Array(4);
      arr[0] = 1.1;              // Init slot 0
      if (overwrite) {
          arr[0] = 3.14;         // Conditionally OVERWRITE same slot 0
      }
      if (escape) {
          return arr;
      }
  }
```
Would the hole value accidentally set in 
```
m_pointerSSA.newDef(variable, block, defaultFor(location));
``` 
or 
```
Node* bottom = defaultFor(location);
m_localMapping.set(location, bottom);
```

hyjorc1 replied:
Basically, I am asking can the placeholder values (`m_bottom/m_empty/m_PNaN`) that are used as SSA initial definitions accidentally end up being materialized into the actual array, even when they shouldn't be?

kmiller68 replied:
No, the reaching def is still either 1.1 or 3.14 and not the hole. I'll add that test case though.

hyjorc1 reviewed (approved):
r=me

Constellation reviewed (approved):
r=me

kmiller68 committed:
Conditional writes are incorrectly handled in Array allocation sinking
https://bugs.webkit.org/show_bug.cgi?id=299956
rdar://161681941

Reviewed by Yusuke Suzuki and Yijia Huang.

The current bottom value in ObjectAllocationSinking is incorrect for arrays.
Unlike with objects, which track conditional stores by passing the active
structure through SSA, arrays can't do this. Instead we should set default value
to the appropriate hole value for the given IndexingShape. To make this work
I had to fix some Phi/Upsilon ResultFormat bugs since they previously assumed
everything would be a JSValue.

Also, add ASSERT to FTL lowering that the Phi/Upsilon formats match. I spent 1/2 a day
trying to understand why I was getting zero, when the issue was those values disagreed
and I was getting the default zero value.

Tests: JSTests/stress/array-allocation-sink-conditional-write-osr.js
       JSTests/stress/array-sink-materialize-conditional-write-argument-value.js
       JSTests/stress/array-sink-materialize-conditional-write.js

Canonical link: https://commits.webkit.org/300888@main
